src/Http/Http/src/Features/QueryFeature.cs

```diff
@@ -100,5 +102,237 @@ public IQueryCollection Query
                 }
             }
         }
+
+        /// <summary>
+        /// Parse a query string into its component key and value parts.
+        /// </summary>
+        /// <param name="queryString">The raw query string value, with or without the leading '?'.</param>
+        /// <returns>A collection of parsed keys and values, null if there are no entries.</returns>
+        [SkipLocalsInit]
+        internal static AdaptiveCapacityDictionary<string, StringValues>? ParseNullableQueryInternal(string? queryString)
+        {
+            if (string.IsNullOrEmpty(queryString) || (queryString.Length == 1 && queryString[0] == '?'))
+            {
+                return null;
+            }
+
+            KvpAccumulator accumulator = new();
+            var queryStringLength = queryString.Length;
+
+            char[]? arryToReturnToPool = null;
+            Span<char> query = (queryStringLength <= 128
+                ? stackalloc char[128]
+                : arryToReturnToPool = ArrayPool<char>.Shared.Rent(queryStringLength)
+            ).Slice(0, queryStringLength);
+
+            queryString.AsSpan().CopyTo(query);
```

Comment:
The only part of this code that requires a copy is the ReplacePlusWithSpaceInPlace calls, right? A string is allocated right after that anyways, so could the two operations be combine then and save the buffer allocation here?

Reply:
Good idea --> b9f61a07ec5965cf7ecf56426268ccb20ee29ab6